kanidmd/src/lib/subscriber.rs

```diff
@@ -0,0 +1,154 @@
+            .insert(KaniLogBuffer::new(/* ctx stuff here */));
+
+        let _ = attrs;
+    }
+
+    fn on_event(&self, event: &Event, ctx: Context<Registry>) {
+        // How can I make it so that this wraps one of `tracing_subscriber`s
+        // preexisting formatting utilities?
+        // I don't want to do the formatting by myself ;(
+        let span = match ctx.lookup_current() {
+            Some(span) => span,
+            _ => {
+                // We're not in any spans, do we still care about the log?
+                // Let's just ignore it for now and short-circuit.
+                return;
+            }
+        };
+
+        // `extensions_mut` returns an `ExtensionsMut`, which is essentially a
+        // wrapping for the `AnyMap` type offered in https://docs.rs/anymap/0.12.1/anymap/
+        let mut ext_mut = span.extensions_mut();
+
+        let logbuf = ext_mut
+            .get_mut::<KaniLogBuffer>()
+            .expect("doesn't have a logbuf even though we initialized one?");
+
+        // TODO: use some formatter to transform this event into a nice string
+        let event_string = format!("{:?}", event);
+
+        logbuf.log(event_string);
+    }
+
+    fn on_exit(&self, id: &Id, ctx: Context<Registry>) {
+        let span = ctx.span(id).expect("how is the span not in ctx?");
+
+        let mut ext_mut = span.extensions_mut();
+
+        let logs = ext_mut
+            // We can take the `KaniLogBuffer` here because we're done with it
+            .remove::<KaniLogBuffer>()
+            .expect("doesn't have a logbuf even though we initialized one?")
+            .dump();
+
+        match span.parent() {
+            Some(parent) => {
+                // There exists a parent span
+                // Write to parent
+                let mut ext_mut = parent.extensions_mut();
+
+                ext_mut
+                    .get_mut::<KaniLogBuffer>()
+                    .expect("has to be here")
+                    .log(logs);
+            }
+            None => {
+                // There is no parent span
+                // Write to `stderr`
+                eprintln!("{}", logs);
+            }
+        }
+    }
+}
+
+impl KaniLogBuffer {
+    pub fn new() -> Self {
+        let mut logbuf = KaniLogBuffer {
+            // TODO: use ctx to get information about what this should be
+            path: "a path".to_string(),
+            buf: vec![],
+        };
+        // TODO: update this to use the ctx to get span information
+        logbuf.log(format!("INFO: {}: OPENED", logbuf.path));
+        logbuf
+    }
+
+    pub fn log(&mut self, event: String) {
+        self.buf.push(event)
+    }
+
+    pub fn dump(mut self) -> String {
+        // TODO: update this to use the ctx to get span information
+        self.log(format!("INFO: {}: CLOSED", self.path));
+        self.buf.join("\n")
+    }
+}
+
+impl KaniSubscriber {
+    pub fn new() -> Self {
+        println!("OOGA NEW SUBSCRIBER");
```

Comment:
Everything here is super temporary, I was just doing an initial implementation to check that I could insert my own arbitrary code that can do stuff when spans start and end.